src/scenic/core/visibility.py

```diff
@@ -0,0 +1,503 @@
+                # distance than the target.
+                object_distances = np.linalg.norm(object_hit_info[0] - np.array(position), axis=1)
+
+                occluded_rays = set()
+
+                for hit_iter in range(len(object_hit_info[0])):
+                    hit_ray = tuple(candidate_ray_list[object_hit_info[1][hit_iter]])
+                    hit_dist = object_distances[hit_iter]
+
+                    # If this ray hit the object earlier than it hit the target, reject the ray.
+                    if hit_dist <= target_dist_map[hit_ray]:
+                        occluded_rays.add(hit_ray)
+
+                candidate_rays = candidate_rays - occluded_rays
+
+                ## DEBUG ##
+                # Show occluded and non occluded rays from this object
+                if debug:
+                    occluded_vertices = [visibleDistance*np.array(vec) + position.coordinates for vec in occluded_rays]
+                    clear_vertices = [visibleDistance*np.array(vec) + position.coordinates for vec in candidate_rays]
+                    vertices = occluded_vertices + clear_vertices
+                    vertices = [position.coordinates] + vertices
+                    lines = [trimesh.path.entities.Line([0,v]) for v in range(1,len(vertices))]
+                    occluded_colors = [(255,0,0,255) for line in occluded_vertices]
+                    clear_colors = [(0,255,0,255) for line in clear_vertices]
+                    colors = occluded_colors + clear_colors
+                    render_scene = trimesh.scene.Scene()
+                    render_scene.add_geometry(trimesh.path.Path3D(entities=lines, vertices=vertices, process=False, colors=colors))
+                    render_scene.add_geometry(target.occupiedSpace.mesh)
+                    render_scene.add_geometry(occ_obj.occupiedSpace.mesh)
+                    render_scene.show()
+
+            if len(candidate_rays) > 0:
+                return True
+
+        # No rays hit the object and are not occluded, so the object is not visible
+        return False
+
+    elif isinstance(target, (Point, OrientedPoint, Vector)):
+        if isinstance(target, (Point, OrientedPoint)):
+            target_loc = target.position
+        else:
+            target_loc = target
+
+        # First check if the distance to the point is less than or equal to the visible distance. If not, the object cannot
+        # be visible.
+        target_distance = position.distanceTo(target_loc)
+        if target_distance > visibleDistance:
+            return False
+
+        # Create the single candidate ray and check that it's within viewAngles.
+        target_vertex = target_loc - position
+        candidate_ray = target_vertex/np.linalg.norm(target_vertex)
+
+        azimuth = np.mod(np.arctan2(candidate_ray[1], candidate_ray[0]) - math.pi/2 + np.pi, 2*np.pi) - np.pi
+        altitude = np.arcsin(candidate_ray[2]/(np.linalg.norm(candidate_ray)))
```

Comment:
You already normalized `candidate_ray` so there's no need for the division here.